sources/freshdesk/freshdesk_client.py

```diff
@@ -0,0 +1,109 @@
+"""Freshdesk source helpers"""
+
+import logging
+import time
+from typing import Any, Iterable, Optional
+
+import pendulum
+from dlt.common.typing import TDataItem
+from dlt.sources.helpers import requests
+
+
+class FreshdeskClient:
+    """
+    Client for making authenticated requests to the Freshdesk API. It incorporates API requests with
+    rate limit and pagination.
+
+    Attributes:
+        api_key (str): The API key used for authenticating requests to the Freshdesk API.
+        domain (str): The Freshdesk domain specific to the user, used in constructing the base URL.
+        base_url (str): The base URL constructed from the domain, targeting the Freshdesk API v2.
+    """
+
+    def __init__(self, api_key: str, domain: str):
+        # Initialize the FreshdeskClient instance with API key and domain.
+        # The API key is used for authentication with the Freshdesk API.
+        # The domain specifies the unique Freshdesk domain of the user.
+
+        # Store the API key provided during initialization.
+        self.api_key = api_key
+        # Store the Freshdesk domain provided during initialization.
+        self.domain = domain
+
+        # Construct the base URL for the API requests.
+        # This URL is formed by appending the domain to the standard Freshdesk API base URL format.
+        # All API requests will use this base URL as their starting point.
+        self.base_url = f"https://{domain}.freshdesk.com/api/v2"
+
+    def _request_with_rate_limit(self, url: str, **kwargs: Any) -> requests.Response:
+        """
+        Handles rate limits in HTTP requests and ensures
+        that the client doesn't exceed the limit set by the server.
+        """
+        while True:
+            try:
+                response = requests.get(url, **kwargs, auth=(self.api_key, "X"))
+                response.raise_for_status()
+                return response
+            except requests.HTTPError as e:
+                if e.response.status_code == 429:
+                    logging.warning("Rate limited. Waiting to retry...")
+                    seconds_to_wait = (
+                        int(e.response.headers["X-Rate-Limit-Reset"])
```

Comment:
you should use key "Retry-After" instead of "X-Rate-Limit-Reset"
Unfortunately, all API have different designs. We should always carefully explore their doc. 
https://developers.freshdesk.com/api/#ratelimit
![image](https://github.com/dlt-hub/verified-sources/assets/20367975/aa3cbc83-5d2c-4b57-ae63-44611402426b)

Reply:
when you write a code you should ALWAYS understand what you are writing, why you using these methods or keys or anything, if you don't know - google, if google doesn't help, ask me :)

Reply:
also, when you catch any error with try-except you should always check if you handle this error properly. run this method so many times until you reach the rate limit

Reply:
also in this case you don't need to subtract pendulum.now().timestamp() I don't remember why we doing it in workable, but for freshdesk it's not needed